File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0047_Permutations_II.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Question_0047_Permutations_II {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null) {
            return res;
        }
        help(nums, 0, res);
        return res;
    }

    private void help(int[] nums, int i, List<List<Integer>> res) {
        if (i >= nums.length) {
            List<Integer> list = new ArrayList<>();
            for (int n : nums) {
                list.add(n);
            }
            res.add(list);
            return;
        }
        Set<Integer> set = new HashSet<>();
        for (int j = i; j < nums.length; j++) {
            if (!set.contains(nums[j])) {
                set.add(nums[j]);
                swap(nums, i, j);
                help(nums, i + 1, res);
                swap(nums, i, j);
            }
        }
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0047_Permutations_II.cc
#include <vector>
#include <set>

using namespace std;

class Solution {
public:
  vector<vector<int>> permuteUnique(vector<int>& nums) {
    vector<vector<int>> res;
    if (nums.empty()) {
      return res;
    }
    help(nums, 0, res);
    return res;
  }

  void help(vector<int>& nums, int i, vector<vector<int>>& res) {
    if (i >= nums.size()) {
      res.push_back(nums);
      return;
    }
    set<int> s;
    for (int j = i; j < nums.size(); j++) {
      if (s.insert(nums[j]).second) {
        swap(nums[i], nums[j]);
        help(nums, i + 1, res);
        swap(nums[i], nums[j]);
      }
    }
  }
};
